crates/autopilot/src/domain/fee/policy.rs

```diff
@@ -84,13 +84,23 @@ impl PriceImprovement {
 }
 
 impl Volume {
-    pub fn apply(&self, order: &boundary::Order) -> Option<domain::fee::Policy> {
+    pub fn apply(
+        &self,
+        order: &boundary::Order,
+        factor_override: Option<FeeFactor>,
+    ) -> Option<domain::fee::Policy> {
         match order.metadata.class {
             boundary::OrderClass::Market => None,
             boundary::OrderClass::Liquidity => None,
-            boundary::OrderClass::Limit => Some(domain::fee::Policy::Volume {
-                factor: self.factor,
-            }),
+            boundary::OrderClass::Limit => {
+                // No volume fee for same-token trades
+                if order.data.buy_token == order.data.sell_token {
+                    return None;
+                }
```

Comment:
Which fee gets applied is now spread across 3 places in the code:
1. base fee
2. fee override
3. this extra exemption

IMO there should be 1 function we can call that already provides the final fee policy taking every override and edge case into account.